lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -974,6 +974,85 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     return StmtDiff(condExpr, ResultRef);
   }
 
+  StmtDiff
+  ReverseModeVisitor::VisitCXXForRangeStmt(const CXXForRangeStmt* FRS) {
+    beginScope(Scope::DeclScope | Scope::ControlScope | Scope::BreakScope |
+               Scope::ContinueScope);
+    beginBlock(direction::reverse);
+
+    LoopCounter loopCounter(*this);
+    const auto* RangeDecl = FRS->getRangeStmt();
+    const auto* BeginDecl = FRS->getBeginStmt();
+    StmtDiff VisitRange = Visit(RangeDecl);
+    StmtDiff VisitBegin = Visit(BeginDecl);
+    Expr* BeginExpr = cast<BinaryOperator>(VisitBegin.getStmt())->getLHS();
+
+    beginBlock(direction::reverse);
+    // Create all declarations needed.
+    Expr* d_BeginDeclRef = m_Variables[cast<DeclRefExpr>(BeginExpr)->getDecl()];
+
+    auto* RangeExpr =
+        cast<DeclRefExpr>(cast<BinaryOperator>(VisitRange.getStmt())->getLHS());
+    auto* BeginDeclRef = cast<DeclRefExpr>(BeginExpr);
+
+    auto* RangeInit = Clone(FRS->getRangeInit());
+    Expr* AssignRange =
+        BuildOp(BO_Assign, RangeExpr, BuildOp(UO_AddrOf, RangeInit));
+    Expr* AssignBegin =
+        BuildOp(BO_Assign, BeginDeclRef, BuildOp(UO_Deref, RangeExpr));
+    addToCurrentBlock(AssignRange);
+    addToCurrentBlock(AssignBegin);
+    const auto* EndDecl = cast<VarDecl>(FRS->getEndStmt()->getSingleDecl());
+
+    Expr* EndInit = cast<BinaryOperator>(EndDecl->getInit())->getRHS();
+    QualType EndType = CloneType(EndDecl->getType());
+    std::string EndName = EndDecl->getNameAsString();
+    Expr* EndAssign = BuildOp(BO_Add, BuildOp(UO_Deref, RangeExpr), EndInit);
+    VarDecl* EndVarDecl =
+        BuildGlobalVarDecl(EndType, EndName, EndAssign, false);
+    DeclStmt* AssignEnd = BuildDeclStmt(EndVarDecl);
+
+    addToCurrentBlock(AssignEnd);
+    DeclRefExpr* EndExpr =
+        BuildDeclRef(cast<VarDecl>(cast<DeclStmt>(AssignEnd)->getSingleDecl()));
+
+    llvm::SaveAndRestore<bool> SaveIsInsideLoop(isInsideLoop);
+    isInsideLoop = true;
+    const VarDecl* LoopVD = FRS->getLoopVariable();
+    Expr* IncBegin = BuildOp(UO_PreInc, BeginDeclRef);
+    Expr* d_IncBegin = BuildOp(UO_PreInc, d_BeginDeclRef);
+
+    Expr* d_DecBegin = BuildOp(UO_PostDec, d_BeginDeclRef);
+    Expr* ForwardCond = BuildOp(BO_NE, BeginDeclRef, EndExpr);
+    // Add item assignment statement to the body.
+    Stmt* bodyClone = Clone(FRS->getBody());
+    CompoundStmt* body = utils::PrependAndCreateCompoundStmt(
+        m_Sema.getASTContext(), bodyClone,
+        BuildDeclStmt(const_cast<VarDecl*>(LoopVD)));
+    StmtDiff BodyDiff =
+        DifferentiateLoopBody(body, loopCounter, nullptr, d_DecBegin,
+                              /*isForLoop=*/true);
+
+    Expr* Inc = BuildOp(BO_Comma, IncBegin, d_IncBegin);
+    Stmt* Forward = new (m_Context)
+        ForStmt(m_Context, nullptr, ForwardCond, nullptr, Inc,
+                BodyDiff.getStmt(), FRS->getForLoc(), noLoc, noLoc);
```

Comment:
```suggestion
                BodyDiff.getStmt(), FRS->getForLoc(), FRS->getBeginLoc(), FRS->getEndLoc());
```

We also need to add comments for the `nullptr`s on the line above.